### ISSUE TITLE
Wazuh with Search Guard

### ISSUE THREAD

**Issue by hoboscientist** [wazuh/wazuh-kibana-app#913]
Panels are not loading in the Wazuh App.
Logs are visible in discover tab.
Is search guard still not supported ?
Is there any configurations necessary for Wazuh Kibana App to work with Search Guard ?

**Comment by jesusgn90** [comment]
Hello @hoboscientist ,

Currently we are researching about SearchGuard and we are planning to fully support it as soon as we can. In the mean time we are building dev environments, configurating instances that use SearchGuard, testing the Wazuh App behaviour and its requirements, etc.

For now, we have a basic configuration files that worked for us but probably they can be improved and we can restrict our roles even more.

I'm assuming you have already installed SearchGuard from your comment and we are probably facing a role problem.

Our Kibana config (/etc/kibana/kibana.yml) looks like this:

```yml
# Use HTTPS instead of HTTP 
elasticsearch.url: "https://localhost:9200"

# Configure the Kibana internal server user
elasticsearch.username: "admin_user"
elasticsearch.password: "admin_pass"

# Disable SSL verification because we use self-signed demo certificates 
elasticsearch.ssl.verificationMode: none

# Whitelist the Search Guard Multi Tenancy Header 
elasticsearch.requestHeadersWhitelist: [ "Authorization", "sgtenant" ]
```

Our Logstash output(/etc/logstash/conf.d/01-wazuh.conf) looks like this:

```js
output {
    elasticsearch {
        hosts => ["localhost:9200"]
        index => "wazuh-alerts-3.x-%{+YYYY.MM.dd}"
        document_type => "wazuh"
        user => logstash
        password => logstash
        ssl => true
        ssl_certificate_verification => false
  }
}
```

You also need to modify Logstash role (<SearchGuard_path>/sgconfig/roles.yml):

```yml
sg_logstash:
  cluster:
    - CLUSTER_MONITOR
    - CLUSTER_COMPOSITE_OPS
    - indices:admin/template/get
    - indices:admin/template/put
  indices:
    'logstash-*':
      '*':
        - CRUD
        - CREATE_INDEX
    '*beat*':
      '*':
        - CRUD
        - CREATE_INDEX
    'wazuh*':
      '*':
        - CRUD
        - CREATE_INDEX
```

To change the passwords, use `<SearchGuard_path>/tools/hash.sh -p "plain_pass"` and replace it in the "hash" field inside the sgconfig/internal_users.yml

Let's apply all the changes:

```
<SearchGuard_path>/tools/sgadmin.sh -cd <SearchGuard_path>/sgconfig -icl -key /etc/elasticsearch/kirk-key.pem -cert /etc/elasticsearch/kirk.pem -cacert /etc/elasticsearch/root-ca.pem
```

Now restart Elastic stack services:

```
systemctl restart logstash
systemctl restart kibana
systemctl restart elasticsearch
```

I hope it helps. As I said, we are currently researching/working on SearchGuard, any feedback from your side is welcome. Thanks in advance.

Regards,
Jesús

**Comment by hoboscientist** [comment]
What do you mean by : 

# Configure the Kibana internal server user
elasticsearch.username: "admin_user"
elasticsearch.password: "admin_pass"

I have configured elasticsearch.username as "kibanaserver", do I need to explicitly provide admin privileges to Kibana ? I am concerned with the security loophole it will introduce in case of Kibana exploit.

**Comment by jesusgn90** [comment]
Hello again @hoboscientist ,

For now, we must use admin privileges for Kibana in order to make the App work as expected. In fact, your are right and we must restrict Kibana server user privileges and Kibana UI user privileges like we are doing with X-Pack security (see https://documentation.wazuh.com/current/user-manual/kibana-app/configure-xpack/index.html for detailes).

Both products work in a similar way, we must define roles, design the users architecture, etc. As I said we are working on it and the idea is to start from high privileges and reducing them while we are testing all the features. 

On the other hand, you should use same version for Kibana and Elasticsearch, there are unexpected behaviour when using different versions. Also we are supporting up to 6.4.x plus Wazuh 3.6.1, you may upgrade in my opinion @hoboscientist !

Stay updated tracking this ticket, we'll publish here news about SearchGuard. Thanks for your patience.

Best regards,
Jesús

**Comment by jesusgn90** [comment]
Hello again @hoboscientist ,

__Just a follow up, not tested at all__

This guide is a first approach to make it work from Wazuh manager to the Wazuh app. Use it under your own risk.

### Instances

This lab environment is composed by four instances: one Wazuh manager, two Elasticsearch nodes and one Kibana. See next picture for details:

![instances](https://user-images.githubusercontent.com/8860227/46597801-e4360d00-cae1-11e8-9455-460c56e36df3.png)

Wazuh manager: 

```yml
filebeat:
 prospectors:
  - type: log
    paths:
     - "/var/ossec/logs/alerts/alerts.json"
    document_type: json
    json.message_key: log
    json.keys_under_root: true
    json.overwrite_keys: true

output:
 logstash:
   # The Logstash hosts
   hosts: ["10.0.0.4:5000","10.0.0.7:5000"]
#   ssl:
#     certificate_authorities: ["/etc/filebeat/logstash.crt"]
```

Kibana: 

```yml
# kibana.yml
elasticsearch.url: "https://10.0.0.4:9200" 
elasticsearch.username: admin 
elasticsearch.password: admin
elasticsearch.ssl.verificationMode: none
elasticsearch.requestHeadersWhitelist: ["Authorization","sgtenant"]
```

Elasticsearch master node:

```yml
# elasticsearch.yml
network.host: 10.0.0.4
cluster.name: "my-cluster"
node.name: "es-node-1"
node.master: true
discovery.zen.ping.unicast.hosts: ["10.0.0.4","10.0.0.7"]
```

```
// 01-wazuh.conf
...
output {
    elasticsearch {
        hosts => ["10.0.0.4:9200"]
        index => "wazuh-alerts-3.x-%{+YYYY.MM.dd}"
        document_type => "wazuh"
        user => logstash
        password => logstash
        ssl => true
        ssl_certificate_verification => false
  }
}
```

Elasticsearch worker node:

```yml
# elasticsearch.yml
network.host: 10.0.0.7
cluster.name: "my-cluster"
node.name: "es-node-2"
node.master: false
discovery.zen.ping.unicast.hosts: ["10.0.0.4","10.0.0.7"]
```

```
// 01-wazuh.conf
...
output {
    elasticsearch {
        hosts => ["10.0.0.7:9200"]
        index => "wazuh-alerts-3.x-%{+YYYY.MM.dd}"
        document_type => "wazuh"
        user => logstash
        password => logstash
        ssl => true
        ssl_certificate_verification => false
  }
}
```

### Installing SearchGuard plugin for Elasticsearch

__Note: Execute same commands for worker node.__

Check the right version here: https://docs.search-guard.com/latest/search-guard-versions

```
com.floragunn:search-guard-6:<elastic_version>-<searchguard_rev>
```

Use the user `elasticsearch` to install the plugin:

```
sudo -u elasticsearch /usr/share/elasticsearch/bin/elasticsearch-plugin install -b com.floragunn:search-guard-6:6.4.1-23.1
```

Install the demo configuration:

```
cd /usr/share/elasticsearch/plugins/search-guard-6/tools/
chmod a+x install_demo_configuration.sh
./install_demo_configuration.sh
Install demo certificates? [y/N] y
Initialize Search Guard? [y/N] y
Enable cluster mode? [y/N] y
```

Restart Elasticsearch service:

```
systemctl restart elasticsearch
```

Check if it's properly installed:

```
curl -k -u admin:admin https://10.0.0.4:9200/_searchguard/authinfo?pretty
{
  "user" : "User [name=admin, roles=[admin], requestedTenant=null]",
  "user_name" : "admin",
  "user_requested_tenant" : null,
  "remote_address" : "10.0.0.4:46378",
  "backend_roles" : [
    "admin"
  ],
  "custom_attribute_names" : [
    "attr.internal.attribute1",
    "attr.internal.attribute2",
    "attr.internal.attribute3"
  ],
  "sg_roles" : [
    "sg_all_access",
    "sg_own_index"
  ],
  "sg_tenants" : {
    "admin_tenant" : true,
    "admin" : true
  },
  "principal" : null,
  "peer_certificates" : "0",
  "sso_logout_url" : null
}
```

Let's create an index with a different naming convention (pattern):

```
// Download the template
curl https://raw.githubusercontent.com/wazuh/wazuh/3.7/extensions/elasticsearch/wazuh-elastic6-template-alerts.json -o /tmp/template.json

// Edit the template pattern and save it
vi template.json

// Save it into Elasticsearch as wazuh2
curl -XPUT -k -u admin:admin 'https://10.0.0.4:9200/_template/wazuh2' -H 'Content-Type: application/json' -d @/tmp/template.json

// Copy existing one index using custom-3.x-*
curl -X POST -u admin:admin -k "https://10.0.0.4:9200/_reindex" -H 'Content-Type: application/json' -d'
{            
  "source": {                             
    "index": "wazuh-alerts-3.x-2018.10.01"
  },       
  "dest": {                         
    "index": "custom-3.x-2018.10.01"
  }
}
'
```

Check for `custom-3.x-*` index existance:
```
curl -k -u admin:admin https://10.0.0.7:9200/_cat/indices?pretty
...
green open custom-3.x-2018.10.01           9juXg-IPS4q7d0CKMgHjKQ 5 1  92 0 424.3kb 204.1kb
green open wazuh-alerts-3.x-2018.10.01     4AN082XxSmSTPnkJ95HHAQ 5 1  92 0 581.3kb 290.6kb
...
```

### Logstash role configuration for SearchGuard

Edit the Logstash role:

```yml
vi /usr/share/elasticsearch/plugins/search-guard-6/sgconfig/sg_roles.yml

sg_logstash:
  cluster:
    - CLUSTER_MONITOR
    - CLUSTER_COMPOSITE_OPS
    - indices:admin/template/get
    - indices:admin/template/put
  indices:
    'logstash-*':
      '*':
        - CRUD
        - CREATE_INDEX
    '*beat*': 
      '*':
        - CRUD
        - CREATE_INDEX
    'wazuh*':
      '*':
        - CRUD
        - CREATE_INDEX
```

If you are using indices like `custom-*`:

```yml
custom*':
  '*':
    - CRUD
    - CREATE_INDEX
```

Apply the changes:

```
/usr/share/elasticsearch/plugins/search-guard-6/tools/sgadmin.sh -cd /usr/share/elasticsearch/plugins/search-guard-6/sgconfig -icl -key /etc/elasticsearch/kirk-key.pem -cert /etc/elasticsearch/kirk.pem -cacert /etc/elasticsearch/root-ca.pem -h 10.0.0.4 -nhnv
```

_Note 1: Specific Elasticsearch host (default is localhost):_

```
-h 10.0.0.4
```

_Note 2: Avoid certificates issue:_

```
-nhnv
```

Restart Elasticsearch and Logstash:

```
systemctl restart elasticsearch
systemctl restart logstash
```

### Kibana SearchGuard plugin

Increase max memory for Node.js, otherwise it fails:

```
export NODE_OPTIONS="--max-old-space-size=3072"
```

Install the plugin as usual (this can be done in Kibana dev too):

```

bin/kibana-plugin install https://search.maven.org/remotecontent?filepath=com/floragunn/search-guard-kibana-plugin/6.4.1-15/search-guard-kibana-plugin-6.4.1-15.zip
```

### Kibana roles

__Note: this procedure must be executed only in the master node, one time.__

Our goal is having a backend user with admin privileges and two UI users. First UI user is able to see `wazuh-alerts-3.x-*`, second UI user is able to see `custom-*`. Both users can are able to see `.wazuh` and `.wazuh-version`.


#### Kibana server user: 

User: `admin`
Pass: `admin`

This is the internal user, not the UI user. This user runs the Kibana server.

#### Kibana UI user for `wazuh-alerts-3.x-*`:

User: `wazuhreadonly01`
Pass: `admin` 

Let's configure it under `sg_internal_users.yml`:

```yml
vi /usr/share/elasticsearch/plugins/search-guard-6/sgconfig/sg_internal_users.yml

wazuhreadonly01:
  hash: $2a$12$VcCDgh2NDk07JGN0rjGbM.Ad41qVR/YFJcgHp0UGns5JDymv..TOG
  roles:
    - wazuhreadonly01_role
```

_Note: Password for `$2a$12$VcCDgh2NDk07JGN0rjGbM.Ad41qVR/YFJcgHp0UGns5JDymv..TOG` hash is `admin`_

Now we are going to configure the role for this user:

```yml
vi /usr/share/elasticsearch/plugins/search-guard-6/sgconfig/sg_roles.yml

sg_wazuh_read_only_01:
  cluster:
    - indices:data/read/mget
    - indices:data/read/msearch
    - indices:data/read/search
    - indices:data/read/field_caps
    - CLUSTER_COMPOSITE_OPS
  indices:
    '?kibana':
      '*':
        - MANAGE
        - INDEX
        - READ
        - DELETE

    'wazuh-alerts*':
      '*':
        - indices:admin/mappings/fields/get
        - indices:admin/validate/query
        - indices:data/read/search
        - indices:data/read/msearch
        - indices:data/read/field_stats
        - indices:data/read/field_caps
        - READ
        - SEARCH
    'wazuh-monitoring*':
      '*':
        - indices:admin/mappings/fields/get
        - indices:admin/validate/query
        - indices:data/read/search
        - indices:data/read/msearch
        - indices:data/read/field_stats
        - indices:data/read/field_caps
        - READ
        - SEARCH
```

Last step is set the role mapping for SearchGuard role and backend roles:

```yml
vi /usr/share/elasticsearch/plugins/search-guard-6/sgconfig/sg_roles_mapping.yml

sg_wazuh_read_only_01:
 backendroles:
    - wazuhreadonly01_role
```

#### Kibana UI user for `custom-*`

User: `wazuhreadonly02`
Pass: `admin`

```yml
vi /usr/share/elasticsearch/plugins/search-guard-6/sgconfig/sg_internal_users.yml

wazuhreadonly02:
  hash: $2a$12$VcCDgh2NDk07JGN0rjGbM.Ad41qVR/YFJcgHp0UGns5JDymv..TOG
  roles:
    - wazuhreadonly02_role
```

```yml
vi /usr/share/elasticsearch/plugins/search-guard-6/sgconfig/sg_roles.yml

sg_wazuh_read_only_02:
  cluster:
    - indices:data/read/mget
    - indices:data/read/msearch
    - indices:data/read/search
    - indices:data/read/field_caps
    - CLUSTER_COMPOSITE_OPS
  indices:
    '?kibana':
      '*':
        - MANAGE
        - INDEX
        - READ
        - DELETE

    'custom-*':
      '*':
        - indices:admin/mappings/fields/get
        - indices:admin/validate/query
        - indices:data/read/search
        - indices:data/read/msearch
        - indices:data/read/field_stats
        - indices:data/read/field_caps
        - READ
        - SEARCH
    'wazuh-monitoring*':
      '*':
        - indices:admin/mappings/fields/get
        - indices:admin/validate/query
        - indices:data/read/search
        - indices:data/read/msearch
        - indices:data/read/field_stats
        - indices:data/read/field_caps
        - READ
        - SEARCH
```

```yml
vi /usr/share/elasticsearch/plugins/search-guard-6/sgconfig/sg_roles_mapping.yml

sg_wazuh_read_only_02:
 backendroles:
    - wazuhreadonly02_role
```

### Apply all the changes

__Note: this procedure must be executed only in the master node, one time.__


```
/usr/share/elasticsearch/plugins/search-guard-6/tools/sgadmin.sh -cd /usr/share/elasticsearch/plugins/search-guard-6/sgconfig -icl -key /etc/elasticsearch/kirk-key.pem -cert /etc/elasticsearch/kirk.pem -cacert /etc/elasticsearch/root-ca.pem -h 10.0.0.4 -nhnv
```

Restart Elasticsearch:

```
systemctl restart elasticsearch
```

### Wazuh Kibana app? 

For now you must use next branch (__it requires Wazuh v3.7.x not released yet and Elastic 6.4.2__):

https://github.com/wazuh/wazuh-kibana-app/tree/search-guard

It has a little modification in order to work properly.

#### What's happening in the Wazuh Kibana App?

1. Our server is running using the `admin` user.
2. Our UI is using the user `wazuhreadonly01` or `wazuhreadonly02` depending on the indices we want to use.

Open Kibana (http://10.0.0.5:5601):

![searchguard-02](https://user-images.githubusercontent.com/8860227/46597884-442cb380-cae2-11e8-8e26-3ba31b328999.png)

#### Looking for forbidden indices (`wazuh-alerts-3.x-*`):

![searchguard-03](https://user-images.githubusercontent.com/8860227/46597873-3a0ab500-cae2-11e8-9bea-3b1b90507cf0.png)

#### Looking for allowed indices (`custom-*`):

![searchguard-04](https://user-images.githubusercontent.com/8860227/46597871-35de9780-cae2-11e8-93c2-2526e52a0bd3.png)

#### Index pattern selector, filtering:

> Only allowed index patterns will be shown in the app (like X-Pack security does).

![searchguard-05](https://user-images.githubusercontent.com/8860227/46597867-31b27a00-cae2-11e8-8f61-ebfe019d7dd6.png)

#### Removing an API entry (it's not allowed for this user):

![searchguard-06](https://user-images.githubusercontent.com/8860227/46597850-23fcf480-cae2-11e8-8716-712d36a4fc8c.png)

-----------

Let us know how it looks for you.

Best regards,
Jesús

**Comment by jesusgn90** [comment]
Hello again @hoboscientist , finally we are planning to support Search Guard. The first iteration on our documentation has been done at https://github.com/wazuh/wazuh-documentation/pull/759, it should work for our current production packages. Next iteration will include index level security and a few other interesting things.

Closing this ticket for now, in any case, you can ask us all you need.

Regards,
Jesús
